Mathlib/Analysis/Asymptotics/Uniform.lean

```diff
@@ -0,0 +1,149 @@
+/-
+Copyright (c) 2024 Lawrence Wu. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Lawrence Wu
+-/
+
+import Mathlib.Analysis.Asymptotics.Theta
+
+/-!
+# Uniform Asymptotics
+
+For a family of functions `f : ι × α → E` and `g : α → E`, we can think of
+`f =O[𝓟 s ×ˢ l] fun (i, x) ↦ g x` as expressing that `f i` is O(g) uniformly on `s`.
+
+This file provides methods for constructing `=O[𝓟 s ×ˢ l]` relations (similarly `Θ`)
+and deriving their consequences.
+-/
+
+open Filter
+
+open Topology
+
+namespace Asymptotics
+
+variable {α ι E F : Type*} {s : Set ι}
+
+section Basic
+
+variable [Norm E] [Norm F] {f : ι × α → E} {g : α → F} {l : Filter α}
+
+/-- If f = O(g) uniformly on `s`, then f_i = O(g) for any `i ∈ s`.` -/
+theorem isBigO_of_isBigOUniformly (h : f =O[𝓟 s ×ˢ l] (g ∘ Prod.snd)) {i : ι} (hi : i ∈ s) :
+    (fun x ↦ f (i, x)) =O[l] g := by
+  obtain ⟨C, hC⟩ := h.bound
+  obtain ⟨t, htl, ht⟩ := hC.exists_mem
+  obtain ⟨u, hu, v, hv, huv⟩ := Filter.mem_prod_iff.mp htl
+  refine isBigO_iff.mpr ⟨C, Filter.eventually_iff_exists_mem.mpr ⟨v, hv, ?_⟩⟩
+  exact fun y hy ↦ ht _ <| huv ⟨hu hi, hy⟩
```

Comment:
I guess I don't really understand the restriction to principal filters here. For example, a slight modification yields:
```lean
/-- If `f = O(g)` eventually with respect to `l' ×ˢ s`, then `f_i = O(g)` eventually in `l'`. -/
theorem foo {l' : Filter ι} (h : f =O[l' ×ˢ l] (g ∘ Prod.snd)) :
    ∀ᶠ i in l', (fun x ↦ f (i, x)) =O[l] g := by
  obtain ⟨C, hC⟩ := h.bound
  obtain ⟨t, htl, ht⟩ := hC.exists_mem
  obtain ⟨u, hu, v, hv, huv⟩ := Filter.mem_prod_iff.mp htl
  filter_upwards [hu] with i hi
  refine isBigO_iff.mpr ⟨C, Filter.eventually_iff_exists_mem.mpr ⟨v, hv, ?_⟩⟩
  exact fun y hy ↦ ht _ <| huv ⟨hi, hy⟩
```
from which you can get:
```lean
/-- If f = O(g) uniformly on `s`, then f_i = O(g) for any `i ∈ s`.` -/
theorem isBigO_of_isBigOUniformly (h : f =O[𝓟 s ×ˢ l] (g ∘ Prod.snd)) {i : ι} (hi : i ∈ s) :
    (fun x ↦ f (i, x)) =O[l] g := by
  revert i
  rw [← eventually_principal]
  simpa using foo h
```

Reply:
Yeah, this PR wasn't in the right generality, so I've reworked it.